core/src/autogluon/core/utils/version_utils.py

```diff
@@ -1,68 +1,46 @@
-import autogluon
-import importlib
 import pkgutil
-import pkg_resources
 import platform
+import re
 import sys
-
 from datetime import datetime
+from importlib.metadata import version, distribution
 
+import autogluon
 from autogluon.common.utils.nvutil import cudaInit, cudaSystemGetNVMLVersion
 from autogluon.common.utils.resource_utils import ResourceManager
 
-# We don't include test dependency here
-autogluon_extras_dict = {
-    'autogluon.core': ('all',),
-    'autogluon.common': (),
-    'autogluon.features': (),
-    'autogluon.timeseries': (),
-    'autogluon.tabular': ('all',),
-}
-
-# This is needed because some module are different in its import name and pip install name
-import_name_dict = {
-    'pillow': 'PIL',
-    'pytorch-lightning': 'pytorch_lightning',
-    'scikit-image': 'skimage',
-    'scikit-learn': 'sklearn',
-    'smart-open': 'smart_open',
-    'timm-clean': 'timm',
-}
-
 
 def _get_autogluon_versions():
     """Retrieve version of all autogluon subpackages and its dependencies"""
     versions = dict()
     for pkg in list(pkgutil.iter_modules(autogluon.__path__, autogluon.__name__ + '.')):
-        if pkg.name == 'autogluon.version':  # autogluon.version will be recognized as a submodule by pkgutil. We don't need it
+        # The following packages will be recognized as a submodule by pkgutil -exclude them.
+        if pkg.name in ['autogluon.version', 'autogluon.setup', 'autogluon._internal_']:
             continue
         try:
-            module = importlib.import_module(pkg.name)
-            versions[pkg.name] = module.__version__
-            versions.update(_get_dependency_versions(pkg.name, autogluon_extras_dict.get(pkg.name, ())))
+            versions[pkg.name] = version(pkg.name)
+            versions.update(_get_dependency_versions(pkg.name))
         except ImportError:
             versions[pkg.name] = None
     return versions
 
 
-def _get_dependency_versions(package, extras=()):
+def _get_dependency_versions(package):
     """Retrieve direct dependency of the given package
 
     Args:
         package (str): name of the package
-        extras (tuple, optional): extras in package dependency. Defaults to ().
     """
-    package = pkg_resources.working_set.by_key[package]
-    dependencies = [str(r.key) for r in package.requires(extras=extras)]
+    # Get all requires for the package
+    dependencies = distribution(package).requires
+    # Filter-out test dependencies
+    dependencies = [req for req in dependencies if not bool(re.search('extra.*test', req))]
+    # keep only package name
+    dependencies = [re.findall('[a-zA-Z0-9_\\-]+', req)[0].strip() for req in dependencies]
```

Comment:
Have you compared the output of this function on mainline with the output in the PR? Ditto for other changes.

Reply:
Yes, it fixed a few issues with missing dependencies:


| main                                               | new                                                 |
|----------------------------------------------------|-----------------------------------------------------|
| INSTALLED VERSIONS                                 | INSTALLED VERSIONS                                  |
| ------------------                                 | ------------------                                  |
| date                   : 2023-05-23                | date                   : 2023-05-23                 |
| time                   : 00:49:04.037355           | time                   : 00:58:15.831350            |
| python                 : 3.9.16.final.0            | python                 : 3.9.16.final.0             |
| OS                     : Linux                     | OS                     : Linux                      |
| OS-release             : 5.15.0-1036-aws           | OS-release             : 5.15.0-1036-aws            |
| Version                : #40~20.04.1-Ubuntu SMP... | Version                : #40~20.04.1-Ubuntu SMP ... |
| machine                : x86_64                    | machine                : x86_64                     |
| processor              : x86_64                    | processor              : x86_64                     |
| num_cores              : 8                         | num_cores              : 8                          |
| cpu_ram_mb             : 31640                     | cpu_ram_mb             : 31640                      |
| cuda version           : None                      | cuda version           : None                       |
| num_gpus               : 0                         | num_gpus               : 0                          |
| gpu_ram_mb             : []                        | gpu_ram_mb             : []                         |
| avail_disk_size_mb     : 905325                    | avail_disk_size_mb     : 898628                     |
|                                                    |                                                     |
| accelerate             : 0.16.0                    | accelerate             : 0.16.0                     |
|                                                    | autogluon              : 0.7.1b20230523             |
| autogluon.common       : 0.7.1b20230523            | autogluon.common       : 0.7.1b20230523             |
| autogluon.core         : 0.7.1b20230523            | autogluon.core         : 0.7.1b20230523             |
| autogluon.eda          : 0.7.1b20230523            | autogluon.eda          : 0.7.1b20230523             |
| autogluon.features     : 0.7.1b20230523            | autogluon.features     : 0.7.1b20230523             |
| autogluon.multimodal   : 0.7.1b20230523            | autogluon.multimodal   : 0.7.1b20230523             |
| autogluon.tabular      : 0.7.1b20230523            | autogluon.tabular      : 0.7.1b20230523             |
| autogluon.timeseries   : 0.7.1b20230523            | autogluon.timeseries   : 0.7.1b20230523             |
| boto3                  : 1.26.120                  | boto3                  : 1.26.120                   |
| catboost               : 1.1.1                     | catboost               : 1.1.1                      |
| defusedxml             : 0.7.1                     | defusedxml             : 0.7.1                      |
| evaluate               : 0.3.0                     | evaluate               : 0.3.0                      |
| fairscale              : 0.4.13                    | fairscale              : 0.4.13                     |
| fastai                 : 2.7.12                    | fastai                 : 2.7.12                     |
| gluonts                : 0.12.8                    | gluonts                : 0.12.8                     |
| hyperopt               : 0.2.7                     | hyperopt               : 0.2.7                      |
|                                                    | imodels                : 1.3.17                     |
| ipython                : None                      | ipython                : 8.12.0                     |
| ipywidgets             : 8.0.6                     | ipywidgets             : 8.0.6                      |
| jinja2                 : 3.1.2                     | jinja2                 : 3.1.2                      |
| joblib                 : 1.2.0                     | joblib                 : 1.2.0                      |
| jsonschema             : 4.17.3                    | jsonschema             : 4.17.3                     |
| lightgbm               : 3.3.5                     | lightgbm               : 3.3.5                      |
| matplotlib             : 3.6.3                     | matplotlib             : 3.6.3                      |
| missingno              : 0.5.2                     | missingno              : 0.5.2                      |
| mlforecast             : 0.7.2                     | mlforecast             : 0.7.2                      |
| networkx               : 2.8.8                     | networkx               : 2.8.8                      |
| nlpaug                 : 1.1.11                    | nlpaug                 : 1.1.11                     |
| nltk                   : 3.8.1                     | nltk                   : 3.8.1                      |
| nptyping               : 2.4.1                     | nptyping               : 2.4.1                      |
| numpy                  : 1.23.5                    | numpy                  : 1.23.5                     |
| omegaconf              : 2.2.3                     | omegaconf              : 2.2.3                      |
|                                                    | onnxruntime-gpu        : 1.13.1                     |
| openmim                : None                      | openmim                : 0.3.7                      |
| pandas                 : 1.5.3                     | pandas                 : 1.5.3                      |
| phik                   : 0.12.3                    | phik                   : 0.12.3                     |
| PIL                    : 9.5.0                     | Pillow                 : 9.5.0                      |
| psutil                 : 5.9.5                     | psutil                 : 5.9.5                      |
| pymupdf                : None                      | PyMuPDF                : 1.21.1                     |
| pyod                   : None                      | pyod                   : 1.0.9                      |
| pytesseract            : 0.3.10                    | pytesseract            : 0.3.10                     |
| pytorch_lightning      : 1.9.5                     | pytorch-lightning      : 1.9.5                      |
| pytorch-metric-learning: None                      | pytorch-metric-learning: 1.7.3                      |
| ray                    : 2.3.1                     | ray                    : 2.3.1                      |
| requests               : 2.28.2                    | requests               : 2.28.2                     |
|                                                    | scikit-image           : 0.19.3                     |
|                                                    | scikit-learn           : 1.1.1                      |
|                                                    | scikit-learn-intelex   : None                       |
| scipy                  : 1.10.1                    | scipy                  : 1.10.1                     |
| seaborn                : 0.12.2                    | seaborn                : 0.12.2                     |
| sentencepiece          : 0.1.99                    | sentencepiece          : 0.1.99                     |
| seqeval                : None                      | seqeval                : 1.2.2                      |
| setuptools             : 65.6.3                    | setuptools             : 65.6.3                     |
| shap                   : 0.41.0                    | shap                   : 0.41.0                     |
| skimage                : 0.19.3                    | skl2onnx               : 1.13                       |
| sklearn                : 1.1.1                     | statsforecast          : 1.4.0                      |
| statsforecast          : 1.4.0                     | statsmodels            : 0.13.5                     |
| statsmodels            : 0.13.5                    | suod                   : 0.0.8                      |
| suod                   : None                      | tensorboard            : 2.13.0                     |
| tensorboard            : 2.13.0                    | text-unidecode         : 1.3                        |
| text-unidecode         : None                      | timm                   : 0.6.13                     |
| timm                   : 0.6.13                    | torch                  : 1.13.1+cpu                 |
| torch                  : 1.13.1+cpu                | torchmetrics           : 0.11.4                     |
| torchmetrics           : 0.11.4                    | torchvision            : 0.14.1+cpu                 |
| torchvision            : 0.14.1+cpu                | tqdm                   : 4.65.0                     |
| tqdm                   : 4.65.0                    | transformers           : 4.26.1                     |
| transformers           : 4.26.1                    | ujson                  : 5.7.0                      |
| ujson                  : 5.7.0                     | vowpalwabbit           : 9.4.0                      |
| xgboost                : 1.7.5                     | xgboost                : 1.7.5                      |
| yellowbrick            : 1.5                       | yellowbrick            : 1.5                        |

Reply:
LGTM, very nice.